domain/cloud/state/state.go

```diff
@@ -73,6 +75,234 @@ func (st *State) ListClouds(ctx context.Context, filter *Filter) ([]cloud.Cloud,
 	return result, errors.Trace(err)
 }
 
+// CloudDefaults provides the currently set cloud defaults for a cloud. If the
+// cloud has no defaults or the cloud does not exist a nil error is returned
+// with an empty defaults map.
+func (st *State) CloudDefaults(ctx context.Context, cloudName string) (map[string]string, error) {
+	defaults := map[string]string{}
+
+	db, err := st.DB()
+	if err != nil {
+		return defaults, errors.Trace(err)
+	}
+
+	stmt := `
+SELECT key, value
+FROM cloud_defaults
+INNER JOIN cloud
+ON cloud_defaults.cloud_uuid = cloud.uuid
+WHERE cloud.name = ?
+`
+
+	return defaults, db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		rows, err := tx.QueryContext(ctx, stmt, cloudName)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		var key, value string
+		for rows.Next() {
+			if err := rows.Scan(&key, &value); err != nil {
+				return errors.Trace(stderrors.Join(err, rows.Close()))
+			}
+			defaults[key] = value
+		}
+		return nil
+	})
+}
+
+// UpdateCloudDefaults is responsible for updating default config values for a
+// cloud. This function will allow the addition and updating of attributes.
+// Attributes can also be removed by keys if they exist for the current cloud.
+func (st *State) UpdateCloudDefaults(
+	ctx context.Context,
+	cloudName string,
+	updateAttrs map[string]string,
+	removeAttrs []string,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	deleteBinds, deleteVals := database.SliceToPlaceholder(removeAttrs)
+	deleteStmt := fmt.Sprintf(`
+DELETE FROM cloud_defaults
+WHERE key IN (%s)
+AND cloud_uuid = (SELECT uuid
+                    FROM cloud
+                    WHERE name = ?)
+`[1:], deleteBinds)
+
+	upsertBinds, upsertVals := database.MapToMultiPlaceholder(updateAttrs)
+	upsertStmt := fmt.Sprintf(`
+CREATE TEMP TABLE cloud_default_dump (
+	key TEXT,
+    value TEXT
+);
+    
+INSERT INTO temp.cloud_default_dump VALUES %s;
+
+INSERT INTO cloud_defaults(cloud_uuid, key, value)
+SELECT (SELECT uuid FROM cloud WHERE name = ?) AS cloud_uuid,
+       key,
+       value
+FROM temp.cloud_default_dump
+-- The WHERE clause here is needed to help sqlite parser with ambiguity.
+WHERE true
+ON CONFLICT(cloud_uuid, key) DO UPDATE
+SET value = excluded.value
+WHERE cloud_uuid = excluded.cloud_uuid
+AND key = excluded.key;
```

Comment:
I'm not sure what the tmp (dump) table gives you, can we just not do `INSERT INTO cloud_defaults(cloud_uuod, key, value) VALUES (?, ?, ?), (?, ?, ?) (?, ?, ?)` with a batch insert?

Reply:
Yeah it's basically a decision between two approaches here. I am not sold on either one in particular.

1. Doing it the way suggested above. We would need to modify our transform functions to take a function that allowed us to inject the `cloud_uuid` at ever 3rd position in the resulting slice or where ever it needs to go. That's fine but a lot of rinse and repeat. Unfortunately there is not way inside of an SQL statement to have a static variable supplied for the insert.

2. This option the one currently done is to dump all the key values into a temp table and then use a select on the insert to bring them back out with the static uuid.

My reason for going with 2 at the moment is let the database handle it all. Saves us have to gin around in code building up the solution and instead we can just give the problem to the DB and say have at it.

Performance wise I don't think there would be any great difference one way or the other it's just going to come down to what style we want to take on etc.